### PR TITLE
Feature get category api and display data to UI, implement order and refresh functions

### PR DESCRIPTION
**Pull request by LaggeR2204**

### CODE REVIEW

File: lib/src/blocs/category/category_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import 'package:bloc/bloc.dart';
+import 'package:equatable/equatable.dart';
+
+import '../../models/category.dart';
+import '../../services/category_service/category_service.dart';
+part './category_event.dart';
+part './category_state.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/category/category_bloc.dart:7]: i suggest you import normally

File: lib/src/blocs/category/category_event.dart [hunk 2]
```diff
@@ -0,0 +1,14 @@
+part of './category_bloc.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/category/category_event.dart:1]: i suggest you import normally

File: lib/src/blocs/category/category_state.dart [hunk 3]
```diff
@@ -0,0 +1,30 @@
+part of './category_bloc.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/category/category_state.dart:1]: i suggest you import normally

File: lib/src/blocs/category/category_event.dart [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  const CategoryEvent();
+
+  @override
+  List<Object> get props => [];
+}
+
+class CategoryRequested extends CategoryEvent {
+  final String? orderState;
+
+  const CategoryRequested({this.orderState});
```
Comment (thanhphamtrung) [inline lib/src/blocs/category/category_event.dart:13]: @OverRide
List get props => [];
add this override

File: lib/src/blocs/category_table_header/category_table_header_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import 'package:equatable/equatable.dart';
+
+import '../../constants/category_orders.dart' as category_orders;
+part './category_table_header_event.dart';
+
+class CategoryTableHeaderBloc extends Bloc<CategoryTableHeaderEvent, String> {
+  CategoryTableHeaderBloc() : super(category_orders.Orders.marketCapDESC) {
+    on<CategoryTableHeaderOrderChanged>((event, emit) {
+      switch (event.columnIndex) {
+        case 1:
```
Comment (thanhphamtrung) [inline lib/src/blocs/category_table_header/category_table_header_bloc.dart:11]: make enum for easier recognize what tab is it

File: lib/src/blocs/category_table_header/category_table_header_event.dart [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  const CategoryTableHeaderEvent();
+
+  @override
+  List<Object> get props => [];
+}
+
+class CategoryTableHeaderOrderChanged extends CategoryTableHeaderEvent {
+  final int columnIndex;
+
+  const CategoryTableHeaderOrderChanged({required this.columnIndex});
```
Comment (thanhphamtrung) [inline lib/src/blocs/category_table_header/category_table_header_event.dart:13]: @OverRide
List get props => [];
add this override

File: lib/src/constants/category_orders.dart [hunk 7]
```diff
@@ -1,8 +1,8 @@
 class Orders {
```
Comment (thanhphamtrung) [inline lib/src/constants/category_orders.dart:1]: you mean orderBy?

File: lib/src/services/category_service/app_category_service.dart [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+import '../../config/app_config.dart';
+import '../../config/constants.dart' as app_constants;
+import '../../constants/constants.dart' as string_constants;
+import '../../constants/category_orders.dart' as category_orders;
+import '../../models/category.dart';
+import '../../models/shortened_category.dart';
+import './category_service.dart';
+
+class AppCategoryService extends CategoryService {
```
Comment (thanhphamtrung) [inline lib/src/services/category_service/app_category_service.dart:13]: implements should be more meaningful than extends

File: lib/src/blocs/category_table_header/category_table_header_event.dart (hunk 6 above)
Comment (LaggeR2204) [reply lib/src/blocs/category_table_header/category_table_header_event.dart:13]: I fixed it

File: lib/src/blocs/category/category_event.dart (hunk 4 above)
Comment (LaggeR2204) [reply lib/src/blocs/category/category_event.dart:13]: I fixed it

File: lib/src/blocs/bottom_navigation_bar/bottom_navigation_bar_event.dart [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  const BottomNavigationBarEvent();
+
+  @override
+  List<Object> get props => [];
+}
+
+class BottomNavigationBarPageChanged extends BottomNavigationBarEvent {
+  final int pageIndex;
+
+  const BottomNavigationBarPageChanged({required this.pageIndex});
```
[the remark this answers is not in the document]
Comment (LaggeR2204) [reply lib/src/blocs/bottom_navigation_bar/bottom_navigation_bar_event.dart:13]: I fixed it

File: lib/src/blocs/category/category_state.dart (hunk 3 above)
Comment (LaggeR2204) [reply lib/src/blocs/category/category_state.dart]: I fix it

File: lib/src/blocs/category/category_event.dart (hunk 2 above)
Comment (LaggeR2204) [reply lib/src/blocs/category/category_event.dart]: I fix it

File: lib/src/blocs/category/category_bloc.dart (hunk 1 above)
Comment (LaggeR2204) [reply lib/src/blocs/category/category_bloc.dart]: I fix it

File: lib/src/constants/category_orders.dart (hunk 7 above)
Comment (LaggeR2204) [reply lib/src/constants/category_orders.dart]: I changed it to OrderBy

File: lib/src/app.dart [hunk 10]
```diff
@@ -1,20 +1,38 @@
...
-      onGenerateRoute: routeController.routePage,
-      initialRoute: route_names.RouteNames.initial,
+    final httpClient = http.Client();
+
+    return MultiBlocProvider(
+      providers: [
+        BlocProvider(create: (context) => BottomNavigationBarBloc()),
+        BlocProvider(
+          create: (context) =>
+              CategoryBloc(service: AppCategoryService(httpClient)),
```
[the remark this answers is not in the document]
Comment (LaggeR2204) [reply lib/src/app.dart:26]: I fix it

File: lib/src/blocs/bottom_navigation_bar/bottom_navigation_bar_bloc.dart [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import 'package:bloc/bloc.dart';
+import 'package:equatable/equatable.dart';
+
+part './bottom_navigation_bar_event.dart';
+
+class BottomNavigationBarBloc extends Bloc<BottomNavigationBarEvent, int> {
+  BottomNavigationBarBloc() : super(0) {
```
[the remark this answers is not in the document]
Comment (LaggeR2204) [reply lib/src/blocs/bottom_navigation_bar/bottom_navigation_bar_bloc.dart]: I added an enum for these pages

File: lib/src/blocs/category_table_header/category_table_header_bloc.dart (hunk 5 above)
Comment (LaggeR2204) [reply lib/src/blocs/category_table_header/category_table_header_bloc.dart]: I made a constant file for it

File: lib/src/services/category_service/app_category_service.dart (hunk 8 above)
Comment (LaggeR2204) [reply lib/src/services/category_service/app_category_service.dart]: I fix it

File: lib/src/blocs/category/category_event.dart [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+  @override
+  List<Object> get props => [];
```
Comment (HaiNguyenPV12) [inline lib/src/blocs/category/category_event.dart:16]: why is `props` not populated?

File: lib/src/blocs/category_table_header/category_table_header_bloc.dart [hunk 13]
```diff
@@ -0,0 +1,38 @@
+import 'package:bloc/bloc.dart';
+import 'package:crypto_app/src/constants/category_columns.dart'
```
Comment (HaiNguyenPV12) [inline lib/src/blocs/category_table_header/category_table_header_bloc.dart:2]: import

**Review by HaiNguyenPV12 — DISMISSED** [review]

**Review by thanhphamtrung — COMMENTED** [review]
import 
empty props Equatable
=> 20

File: lib/src/blocs/category/category_event.dart (hunk 12 above)
Comment (LaggeR2204) [reply lib/src/blocs/category/category_event.dart]: I have populate the prop, but I'm not sure if it is correct, can you please check again

File: lib/src/blocs/category_table_header/category_table_header_bloc.dart (hunk 13 above)
Comment (LaggeR2204) [reply lib/src/blocs/category_table_header/category_table_header_bloc.dart]: I fixed it, sr about this mistake, my 'import' extension has some problems and I didn't check carefully

**Review by thanhphamtrung — DISMISSED** [review]

**Review by thanhphamtrung — APPROVED** [review]

**Pull request merged**